firedrake/interpolation.py

```diff
@@ -124,31 +128,406 @@ class Interpolator(object):
+                else:
+                    expr_subfunctions = self.expr.subfunctions
+                if len(expr_subfunctions) != len(V_dest.subfunctions):
+                    raise NotImplementedError(
+                        "Can't interpolate from a non-mixed function space into a mixed function space."
+                    )
+                for input_sub_func, target_sub_func in zip(
+                    expr_subfunctions, V_dest.subfunctions
+                ):
+                    sub_interpolator = CrossMeshInterpolator(
+                        input_sub_func,
+                        target_sub_func,
+                        subset=subset,
+                        freeze_expr=freeze_expr,
+                        access=access,
+                        bcs=bcs,
+                        allow_missing_points=allow_missing_points,
+                    )
+                    self.sub_interpolators.append(sub_interpolator)
+                return
+        V_dest_vec = firedrake.VectorFunctionSpace(m_dest, ufl_scalar_element)
+        f_dest_node_coords = interpolate(m_dest.coordinates, V_dest_vec)
+        dest_node_coords = f_dest_node_coords.dat.data_ro
+        self.vom_dest_node_coords_in_m_src = firedrake.VertexOnlyMesh(
+            m_src,
+            dest_node_coords,
+            redundant=False,
+            missing_points_behaviour=missing_points_behaviour,
+        )
+        # vom_dest_node_coords_in_m_src uses the parallel decomposition of
+        # the global node coordinates of V_dest in the SOURCE mesh (m_src).
+        # I first point evaluate my expression at these locations, giving a
+        # P0DG function on the VOM. As described in the manual, this is an
+        # interpolation operation.
+        shape = V_dest.ufl_element().value_shape()
+        if len(shape) == 0:
+            fs_type = firedrake.FunctionSpace
+        elif len(shape) == 1:
+            fs_type = firedrake.VectorFunctionSpace
+        else:
+            fs_type = partial(firedrake.TensorFunctionSpace, shape=shape)
+        P0DG_vom = fs_type(self.vom_dest_node_coords_in_m_src, "DG", 0)
+        self.point_eval_interpolator = Interpolator(self.expr, P0DG_vom)
+        # The parallel decomposition of the nodes of V_dest in the DESTINATION
+        # mesh (m_dest) is retrieved using the input_ordering attribute of the
+        # VOM. This again is an interpolation operation, which, under the hood
+        # is a PETSc SF reduce.
+        P0DG_vom_i_o = fs_type(
+            self.vom_dest_node_coords_in_m_src.input_ordering, "DG", 0
+        )
+        self.to_input_ordering_interpolator = Interpolator(
+            firedrake.TrialFunction(P0DG_vom), P0DG_vom_i_o
+        )
+        # The P0DG function outputted by the above interpolation has the
+        # correct parallel decomposition for the nodes of V_dest in m_dest so
+        # we can safely assign the dat values. This is all done in the actual
+        # interpolation method below.
```

Comment:
Should any of this have annotation turned off?

Reply:
https://github.com/firedrakeproject/firedrake/pull/3087